### PR TITLE
Ena 6648 embl fasta reader wrapper

### PR DESCRIPTION
**Pull request by IvaTutis**
- added a wrapper for fastaReader
- added as starting input parameters the list of accessionIds and the file location

### CODE REVIEW

**Review by Rajkumar-D — COMMENTED** [review]
commented

File: src/main/java/uk/ac/ebi/embl/gff3tools/fasta/JsonHeaderFastaReader.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        }
+        for (int i = 0; i < orderedSubmissionIds.size(); i++) {
+            var submissionId = orderedSubmissionIds.get(i);
+            var accessionId = orderedAccessionIds.get(i);
+            accessionIdToSubmissionId.put(accessionId, submissionId);
+            submissionIdToAccessionId.put(submissionId, accessionId);
+        }
+    }
+
+    public String getSubmissionIdByAccessionId(String accessionId)
```
Comment (Istar-Eldritch) [inline src/main/java/uk/ac/ebi/embl/gff3tools/fasta/JsonHeaderFastaReader.java]: The methods `getSubmissionIdByAccessionId` and `getAccessionIdBySubmissionId` declare `throws FastaHeaderParserException, FastaFileException` but never actually throw these exceptions. Consider removing the unnecessary throws declarations.

File: src/main/java/uk/ac/ebi/embl/gff3tools/fasta/headerutils/JsonHeaderParser.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import uk.ac.ebi.embl.gff3tools.exception.FastaHeaderParserException;
+
+public class JsonHeaderParser {
+    private static final ObjectMapper MAPPER = JsonMapper.builder()
+            .enable(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES)
+            .build();
+    private static final int MAX_HEADER_LENGTH = 4096;
+
+    public ParsedHeader parse(String headerLine) throws FastaHeaderParserException {
+
```
Comment (Istar-Eldritch) [inline src/main/java/uk/ac/ebi/embl/gff3tools/fasta/headerutils/JsonHeaderParser.java]: Consider adding null check at the start of the method:
```java
if (headerLine == null) {
    throw new FastaHeaderParserException("Header line cannot be null");
}
```

File: src/main/java/uk/ac/ebi/embl/gff3tools/fasta/headerutils/FastaHeader.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+package uk.ac.ebi.embl.gff3tools.fasta.headerutils;
+
+import com.fasterxml.jackson.annotation.JsonAlias;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class FastaHeader {
```
Comment (Istar-Eldritch) [inline src/main/java/uk/ac/ebi/embl/gff3tools/fasta/headerutils/FastaHeader.java]: The comments indicate `description`, `molecule_type`, and `topology` are mandatory fields, but there's no validation enforcing this. Consider adding validation in `JsonHeaderParser.parseHeaderJson()` to check these fields are present and non-null after parsing.

File: src/test/java/uk/ac/ebi/embl/gff3tools/fasta/JsonHeaderFastaReaderTest.java [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        File fasta = TestUtils.getResourceFile("./fasta/fasta_good_example.txt");
+        List<String> accessionIds = List.of("acc1", "acc2", "acc3");
+
+        try (JsonHeaderFastaReader service = new JsonHeaderFastaReader(fasta)) {
+            service.setAccessionIds(accessionIds);
+            assertEquals("ID1", service.getSubmissionIdByAccessionId("acc1"));
+            assertEquals("ID2", service.getSubmissionIdByAccessionId("acc2"));
+            assertEquals("ID3", service.getSubmissionIdByAccessionId("acc3"));
+            assertNull(service.getSubmissionIdByAccessionId("hehe"));
+            assertEquals("acc1", service.getAccessionIdBySubmissionId("ID1"));
```
Comment (Istar-Eldritch) [inline src/test/java/uk/ac/ebi/embl/gff3tools/fasta/JsonHeaderFastaReaderTest.java]: Consider adding a test for duplicate submission IDs in the FASTA file to verify the expected behavior (either error or defined handling).

File: src/main/java/uk/ac/ebi/embl/gff3tools/fasta/JsonHeaderFastaReader.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        if (fastaReader != null) {
+            fastaReader.close();
+        }
+    }
+
+    private void parseData() throws FastaHeaderParserException, FastaFileException {
+        var entries = fastaReader.getFastaEntries();
+        for (int i = 0; i < entries.size(); i++) {
+            var entry = entries.get(i);
+            // parse headers
```
[the remark this answers is not in the document]
Comment (IvaTutis) [reply src/main/java/uk/ac/ebi/embl/gff3tools/fasta/JsonHeaderFastaReader.java]: valid point about the IDs, I'll add a test for that. The validations were insofar imagined to be done with the validation engine the same way the gff3 validations are done. List here https://docs.google.com/spreadsheets/d/1WOoOg-k2vrWpiH4T2NHIcJHYSSH9inOy-0AwPBNPlTI/edit?gid=523873719#gid=523873719

**Review by Rajkumar-D — APPROVED** [review]
